sklearn/utils/validation.py

```diff
@@ -1588,3 +1588,31 @@ def _check_fit_params(X, fit_params, indices=None):
             )
 
     return fit_params_validated
+
+
+def _get_feature_names(X):
+    """Get feature names from X.
+
+    Support for other array containers should place its implementation here.
+
+    Parameters
+    ----------
+    X : dataframe of shape (n_samples, n_features)
+        Array container to extract feature names.
+
+        - pandas DataFrame : The columns will be considered to be feature
+          names.
+        - All other array containers will return `None`.
+
+    Returns
+    -------
+    names: Sequence of str or None
+        Column names of `X`. Unrecognized array containers will return `None`.
+    """
+    if hasattr(X, "iloc"):
+        with suppress(ImportError):
+            # check explicitly for pandas
+            import pandas as pd
+
+            if isinstance(X, pd.DataFrame):
+                return X.columns
```

Comment:
what does this return with hierarchical column indices?

Reply:
Ah, I guess we have to be strict at first and only allow non-hierarchical column indices, which means some validation here.

Reply:
Yes, and at least warn on integer column names please? :D

Reply:
> Yes, and at least warn on integer column names please? :D

I am thinking of supporting integer column names because one of our Vectorizers output them:

```python
from sklearn.feature_extraction import DictVectorizer
v = DictVectorizer(sparse=False)
D = [{1: 1, 2: 2}, {2: 3, 3: 1}]
X = v.fit(D)
v.get_feature_names()
# [1, 2, 3]
```